File: app/rule_cache.py
"""In-process rule cache (arch §4.1, AC-34, AC-38, AC-49).

A ``dict[token] -> CompiledEndpoint`` that powers the <10ms fast path: the
interceptor reads the endpoint config + the **pre-sorted, pre-compiled** rule list
from memory — it opens **no per-request DB connection** on the matched path.

  * ``get(token)``        — return the cached :class:`CompiledEndpoint`, loading
                            from SQLite on a cold miss (the only DB read, amortized).
  * ``invalidate(token)`` — drop the cache entry; the next ``get`` reloads. Driven
                            by the ``cfg:<token>`` pub/sub signal on any management
                            write (``app/pubsub.py``) and called directly in-process
                            for the single-instance case (AC-34).

Missing tokens are cached negatively for a short TTL so a flood of unknown-token
hits cannot hammer SQLite, while a freshly-created endpoint still appears promptly
(the negative entry is invalidated by ``cfg`` on create, and expires anyway).

Redis is **not** on the match read path (AC-49): a Redis outage never affects
matching — only state-gated rules degrade (fail-closed, evaluated in the matcher).
"""

from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional

import aiosqlite

from config import DATABASE_PATH
from app.interceptor.matcher import CompiledRule, compile_rule

logger = logging.getLogger("hookbox.rule_cache")

# How long a negative (token-not-found) entry stays cached.
_NEG_TTL_SECONDS = 5.0


@dataclass
class CompiledEndpoint:
    """The hot, in-memory view of an endpoint: config + compiled, ordered rules."""

    token: str
    owner_id: str
    auto_crud: bool
    target_url: Optional[str]
    default_mode: str
    latency_ms: int
    rate_limit_per_min: int
    chaos_pct: int
    cors_enabled: bool
    rules: List[CompiledRule] = field(default_factory=list)

    @property
    def any_rule_gates_on_state(self) -> bool:
        return any(r.enabled and r.gates_on_state for r in self.rules)


class RuleCache:
    def __init__(self, db_path=DATABASE_PATH) -> None:
        self._db_path = db_path
        self._cache: Dict[str, CompiledEndpoint] = {}
        self._negative: Dict[str, float] = {}  # token -> expiry monotonic ts

    def invalidate(self, token: str) -> None:
        """Drop any cached (positive or negative) entry for ``token`` (AC-34)."""
        self._cache.pop(token, None)
        self._negative.pop(token, None)

    def invalidate_all(self) -> None:
        self._cache.clear()
        self._negative.clear()

    def peek(self, token: str) -> Optional[CompiledEndpoint]:
        """Non-loading lookup (returns the cached entry or None)."""
        return self._cache.get(token)

    async def get(self, token: str) -> Optional[CompiledEndpoint]:
        """Return the compiled endpoint for ``token`` (cold-loads from SQLite).

        Returns ``None`` if the token does not exist (cached negatively for a few
        seconds). Never raises on a normal miss.
        """
        hit = self._cache.get(token)
        if hit is not None:
            return hit

        neg_exp = self._negative.get(token)
        if neg_exp is not None:
            if neg_exp > time.monotonic():
                return None
            self._negative.pop(token, None)

        loaded = await self._load(token)
        if loaded is None:
            self._negative[token] = time.monotonic() + _NEG_TTL_SECONDS
            return None
        self._cache[token] = loaded
        return loaded

    async def _load(self, token: str) -> Optional[CompiledEndpoint]:
        try:
            async with aiosqlite.connect(self._db_path) as db:
                db.row_factory = aiosqlite.Row
                cur = await db.execute(
                    "SELECT * FROM endpoints WHERE token = ?", (token,)
                )
                ep = await cur.fetchone()
                if ep is None:
                    return None
                cur = await db.execute(
                    "SELECT * FROM mock_rules WHERE token = ? ORDER BY priority, id",
                    (token,),
                )
                rule_rows = await cur.fetchall()
        except Exception:  # noqa: BLE001 - a DB hiccup must not crash the fast path
            logger.exception("rule_cache load failed for %s", token)
            return None

        rules: List[CompiledRule] = []
        for r in rule_rows:
            try:
                rules.append(
                    compile_rule(
                        {
                            "id": r["id"],
                            "priority": r["priority"],
                            "enabled": bool(r["enabled"]),
                            "match": json.loads(r["match_json"] or "{}"),
                            "response": json.loads(r["response_json"] or "{}"),
                            "state_writes": json.loads(r["state_writes_json"] or "[]"),
                            "latency_ms": r["latency_ms"],
                            "rate_limit_per_min": r["rate_limit_per_min"],
                            "webhook_action": json.loads(r["webhook_json"]) if r["webhook_json"] else None,
                        }
                    )
                )
            except Exception:  # noqa: BLE001 - skip a single malformed rule, keep the rest
                logger.exception("skipping malformed rule id=%s on %s", r["id"], token)

        return CompiledEndpoint(
            token=ep["token"],
            owner_id=ep["owner_id"],
            auto_crud=bool(ep["auto_crud"]),
            target_url=ep["target_url"],
            default_mode=ep["default_mode"],
            latency_ms=ep["latency_ms"],
            rate_limit_per_min=ep["rate_limit_per_min"],
            chaos_pct=ep["chaos_pct"],
            cors_enabled=bool(ep["cors_enabled"]),
            rules=rules,
        )


# Module-level singleton (referenced by app/routes/api.py invalidation + pubsub).
rule_cache = RuleCache()

File: backend/src/rule_cache.rs
//! Compiled-endpoint cache — PORT of `app/rule_cache.py` (AC-11/15/38, AC-S18).
//!
//! On the P1 fast path the endpoint config + compiled rules come from this
//! in-process cache: it cold-loads an endpoint once from SQLite, compiles its
//! rules (sorted by priority,id), and serves subsequent requests with NO
//! per-request DB read on a match. Any management write calls `invalidate(token)`
//! so the next P1 request cold-loads fresh config. A tombstoned endpoint
//! (`gone_at` non-null) is cached as `Gone` so 410 is served without a DB read.

use std::sync::Arc;

use dashmap::DashMap;
use sqlx::{Row, SqlitePool};

use crate::interceptor::matcher::{compile_rule, CompiledRule};
use crate::models::{MatchCriteria, ResponseSpec, StateWrite, WebhookAction};

/// A compiled endpoint ready for the fast path.
#[derive(Clone)]
pub struct CompiledEndpoint {
    pub token: String,
    pub auto_crud: bool,
    pub target_url: Option<String>,
    pub default_mode: String, // "mock_404" | "echo"
    pub latency_ms: i64,
    pub rate_limit_per_min: i64,
    pub chaos_pct: i64,
    pub chaos_mode: String, // "error" | "dropout"
    pub cors_enabled: bool,
    pub rules: Arc<Vec<CompiledRule>>,
    pub any_rule_gates_on_state: bool,
}

/// What the cache resolves a token to.
#[derive(Clone)]
pub enum Resolved {
    /// Live endpoint with compiled rules.
    Live(Arc<CompiledEndpoint>),
    /// Tombstoned (deleted/expired) — serve 410 without a DB read.
    Gone,
}

pub struct RuleCache {
    entries: DashMap<String, Resolved>,
}

impl Default for RuleCache {
    fn default() -> Self {
        RuleCache {
            entries: DashMap::new(),
        }
    }
}

impl RuleCache {
    pub fn new() -> Self {
        RuleCache::default()
    }

    /// Drop the cached entry for a token after any management write.
    pub fn invalidate(&self, token: &str) {
        self.entries.remove(token);
    }

    /// Resolve a token: serve from cache, else cold-load from SQLite. Returns
    /// `None` for a genuinely unknown endpoint (no row at all → 404).
    pub async fn get(
        &self,
        pool: &SqlitePool,
        token: &str,
    ) -> Result<Option<Resolved>, sqlx::Error> {
        if let Some(hit) = self.entries.get(token) {
            return Ok(Some(hit.clone()));
        }
        let row = sqlx::query("SELECT * FROM endpoints WHERE token = ?")
            .bind(token)
            .fetch_optional(pool)
            .await?;
        let row = match row {
            Some(r) => r,
            None => return Ok(None), // unknown -> 404 (not cached)
        };
        let gone_at: Option<String> = row.get("gone_at");
        if gone_at.is_some() {
            self.entries.insert(token.to_string(), Resolved::Gone);
            return Ok(Some(Resolved::Gone));
        }

        let rule_rows =
            sqlx::query("SELECT * FROM mock_rules WHERE token = ? ORDER BY priority, id")
                .bind(token)
                .fetch_all(pool)
                .await?;
        let mut rules = Vec::with_capacity(rule_rows.len());
        let mut any_state = false;
        for r in &rule_rows {
            let match_json: String = r.get("match_json");
            let response_json: String = r.get("response_json");
            let writes_json: String = r.get("state_writes_json");
            let webhook_json: Option<String> = r.get("webhook_json");
            let mc: MatchCriteria = serde_json::from_str(&match_json).unwrap_or_default();
            let rsp: ResponseSpec = serde_json::from_str(&response_json).unwrap_or_default();
            let writes: Vec<StateWrite> = serde_json::from_str(&writes_json).unwrap_or_default();
            let webhook: Option<WebhookAction> =
                webhook_json.and_then(|w| serde_json::from_str(&w).ok());
            let compiled = compile_rule(
                r.get("id"),
                r.get("priority"),
                r.get::<i64, _>("enabled") != 0,
                &mc,
                rsp,
                writes,
                r.get("latency_ms"),
                r.get("rate_limit_per_min"),
                r.get("chaos_mode"),
                webhook,
            );
            if compiled.gates_on_state() {
                any_state = true;
            }
            rules.push(compiled);
        }

        let ep = Arc::new(CompiledEndpoint {
            token: token.to_string(),
            auto_crud: row.get::<i64, _>("auto_crud") != 0,
            target_url: row.get("target_url"),
            default_mode: row.get("default_mode"),
            latency_ms: row.get("latency_ms"),
            rate_limit_per_min: row.get("rate_limit_per_min"),
            chaos_pct: row.get("chaos_pct"),
            chaos_mode: row.get("chaos_mode"),
            cors_enabled: row.get::<i64, _>("cors_enabled") != 0,
            rules: Arc::new(rules),
            any_rule_gates_on_state: any_state,
        });
        let resolved = Resolved::Live(ep);
        self.entries.insert(token.to_string(), resolved.clone());
        Ok(Some(resolved))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db;

    async fn pool() -> SqlitePool {
        let p = db::pool(":memory:").await.unwrap();
        db::migrate(&p).await.unwrap();
        sqlx::query("INSERT INTO owners (owner_id, email, secret_hash) VALUES ('o','e','h')")
            .execute(&p)
            .await
            .unwrap();
        p
    }

    #[tokio::test]
    async fn cold_load_compile_and_invalidate() {
        let p = pool().await;
        sqlx::query("INSERT INTO endpoints (token, owner_id, auto_crud) VALUES ('tok','o',1)")
            .execute(&p)
            .await
            .unwrap();
        sqlx::query(
            "INSERT INTO mock_rules (token, priority, enabled, match_json, response_json, state_writes_json)
             VALUES ('tok', 50, 1, '{\"method\":\"GET\",\"path\":\"/*\"}', '{\"status_code\":200}', '[]')",
        ).execute(&p).await.unwrap();
        let cache = RuleCache::new();
        let r = cache.get(&p, "tok").await.unwrap().unwrap();
        match r {
            Resolved::Live(ep) => {
                assert!(ep.auto_crud);
                assert_eq!(ep.rules.len(), 1);
                assert_eq!(ep.rules[0].priority, 50);
            }
            _ => panic!("expected live"),
        }
        // unknown -> None
        assert!(cache.get(&p, "nope").await.unwrap().is_none());
        // tombstone -> Gone
        sqlx::query("UPDATE endpoints SET gone_at = datetime('now') WHERE token='tok'")
            .execute(&p)
            .await
            .unwrap();
        cache.invalidate("tok");
        assert!(matches!(
            cache.get(&p, "tok").await.unwrap().unwrap(),
            Resolved::Gone
        ));
    }
}
